src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs

```diff
@@ -146,6 +150,38 @@ internal MsQuicStream(MsQuicConnection.State connectionState, SafeMsQuicStreamHa
             }
         }
 
+        internal static async ValueTask<MsQuicStream> CreateOutbound(MsQuicConnection.State connectionState, QUIC_STREAM_OPEN_FLAGS flags, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+            var stream = new MsQuicStream(connectionState, flags);
+            State state = stream._state;
+
+            try
+            {
+                Debug.Assert(!Monitor.IsEntered(state));
+
+                cancellationToken.ThrowIfCancellationRequested();
+                using CancellationTokenRegistration registration = cancellationToken.UnsafeRegister(static (s, token) =>
+                {
+                    ((State)s!).StartCompletionSource.TrySetException(new OperationCanceledException(token));
+                }, state);
+
+                // Fire of start of the stream
+                uint status = MsQuicApi.Api.StreamStartDelegate(state.Handle, QUIC_STREAM_START_FLAGS.SHUTDOWN_ON_FAIL | QUIC_STREAM_START_FLAGS.INDICATE_PEER_ACCEPT);
+                QuicExceptionHelpers.ThrowIfFailed(status, "Could not start stream.");
+
+                // wait unit start completes.
+                await state.StartCompletionSource.Task.ConfigureAwait(false);
```

Comment:
We can't because types don't match (typeless (Value)Task vs `ValueTask<MsQuicStream>`), also, in case the start fails due to an exception (e.g. connection is aborted while we wait for the peer to increase the stream limits), we need to dispose the stream instance correctly.

Reply:
I don't follow, how will the connection take care of that?

if we don't dispose the stream here, then the finalizer would have to take care of that, otherwise the MsQuic stream handle will not get closed because we wait until both dispose and shutdown notification happen before MsQuicConnection.State.Cleanup is called.